libraries/botframework-connector/src/auth/appCredentials.ts

```diff
@@ -0,0 +1,157 @@
+    }
+
+    private static isTrustedUrl(uri: string): boolean {
+        const expiration: Date = AppCredentials.trustedHostNames.get(uri);
+        if (expiration) {
+            // check if the trusted service url is still valid
+            return expiration.getTime() > (Date.now() - 300000); // 5 Minutes
+        }
+
+        return false;
+    }
+
+    public async signRequest(webResource: msrest.WebResource): Promise<msrest.WebResource> {
+        if (this.shouldSetToken(webResource)) {
+            const token: string = await this.getToken();
+
+            return new msrest.TokenCredentials(token).signRequest(webResource);
+        }
+
+        return webResource;
+    }
+
+    public async getToken(forceRefresh: boolean = false): Promise<string> {
+        if (!forceRefresh) {
+            // check the global cache for the token. If we have it, and it's valid, we're done.
+            const oAuthToken: adal.TokenResponse = AppCredentials.cache.get(this.tokenCacheKey);
+            if (oAuthToken) {
+                // we have the token. Is it valid?
+                if (oAuthToken.expirationTime > Date.now()) {
+                    return oAuthToken.accessToken;
+                }
+            }
+        }
+
+        // We need to refresh the token, because:
+        // 1. The user requested it via the forceRefresh parameter
+        // 2. We have it, but it's expired
+        // 3. We don't have it in the cache.
+        const res: adal.TokenResponse = await this.refreshToken();
+        this.refreshingToken = null;
+
+        if (res && res.accessToken) {          
+            // `res` is equalivent to the results from the cached promise `this.refreshingToken`.
+            // Because the promise has been cached, we need to see if the body has been read.
+            // If the body has not been read yet, we can call res.json() to get the access_token.
+            // If the body has been read, the OAuthResponse for that call should have been cached already,
+            // in which case we can return the cache from there. If a cached OAuthResponse does not exist,
+            // call getToken() again to retry the authentication process.
+
+            // Subtract 5 minutes from expires_in so they'll we'll get a
+            // new token before it expires.
+            res.expirationTime = Date.now() + (res.expiresIn * 1000) - 300000;
+            AppCredentials.cache.set(this.tokenCacheKey, res);
+            return res.accessToken;
+        } else {
+            throw new Error('Authentication: No response or error received from ADAL.');
+        }
+
+    }
+
+    protected abstract async refreshToken(): Promise<adal.TokenResponse>;
+
+    private shouldSetToken(webResource: msrest.WebResource): boolean {
+        return AppCredentials.isTrustedServiceUrl(webResource.url);
+    }
+}
+
+/**
+ * Member variables to this class follow the RFC Naming conventions, rather than C# naming conventions.
+ */
+interface OAuthResponse {
```

Comment:
We are not using this anymore, and it doesn't seem to be exported. Are we safe to remove?

Reply:
I was just about to leave a comment for this! Please remove it since it's not being used any more 👍